File: Ejemplos de patrones/Creacionales/Prototype/Java/Ejemplo1.java
// === Ejemplo 1: Prototype básico ===
class Personaje implements Cloneable {
    String nombre;
    String raza;

    public Personaje(String nombre, String raza) {
        this.nombre = nombre;
        this.raza = raza;
    }

    public Personaje clone() throws CloneNotSupportedException {
        return (Personaje) super.clone(); // clon superficial
    }
}

public class Main {
    public static void main(String[] args) throws Exception {
        Personaje p1 = new Personaje("Aragorn", "Humano");
        Personaje p2 = p1.clone();
        System.out.println(p1 == p2); // false
    }
}

File: Ejemplos de patrones/Creacionales/Prototype/Python/Ejemplo1.py
import copy  # importa el módulo copy para realizar copias superficiales y profundas

class Personaje:  # define la clase Personaje
    def __init__(self, nombre, raza):  # constructor que recibe nombre y raza
        self.nombre = nombre  # asigna el nombre al atributo
        self.raza = raza  # asigna la raza al atributo

    def clone(self):  # método que crea un clon superficial
        return copy.copy(self)  # devuelve una copia superficial del objeto

# --- Uso ---
p1 = Personaje("Aragorn", "Humano")  # crea el personaje original
p2 = p1.clone()  # clona el personaje usando copy.copy (shallow)
print(p1 is p2)  # muestra False: p1 y p2 son objetos distintos
print(p1.nombre, p2.nombre)  # imprime los nombres (mismo valor inicial)
